### ISSUE TITLE
Maintain Attestations who reference unknown blocks

### ISSUE THREAD

**Issue by AgeManning** [sigp/lighthouse#635]
When receiving an attestation that references a block that we don't know of, this will trigger a block search. 
Currently, the attestation gets dropped. The preferred logic is to maintain the attestation until the block search either completes or fails and then should be attempted to be reprocessed.

**Comment by AgeManning** [comment]
I think this is becoming increasingly important. 

We are missing some attestations in a race condition in block processing and attestation processing. See for example:
```
 287584 Dec 04 07:10:48.373 DEBG Successfully processed gossip block     root: 0x4973567e5cd7d5001e19b1af609dccc1a47d0aa51cd87ff48087c7ccae6d5ec7, slot: 16852, graffiti: 🍦 C.R.E.A.M. @ TW 🇹🇼, service: beacon
 287585 Dec 04 07:10:48.373 INFO New block received                      hash: 0x4973…5ec7, slot: 16852
 287586 Dec 04 07:10:48.389 DEBG Attestation for unknown block           block: 0x4973…5ec7, peer_id: 16Uiu2HAmJBYJfm6wvBoyFa1cu5ydtdTE61juf4fa2kNCoPV4JdJF  
```

We ideally want to keep a cache of attestations for unknown blocks and try and reprocess them if they match a block we have just processed. Those that expire in the cache (i.e no block was ever found for them) the peers that sent them should get penalized.

**Comment by divagant-martian** [comment]
Interested in working on this but I've been told @blacktemplar might be interested too?

**Comment by blacktemplar** [comment]
Didn't start to work on that yet, so go for it :). Just a small thought about the cache: Would we wait until either the block is known or the cache expires before reporting the validity of the message to gossipsub? Currently gossipsub expects validation to be "fast", but I guess instead of ignoring the message it is always better to just wait and maybe process it later.

**Comment by AgeManning** [comment]
Yeah, currently we ignore it. In principle the block-lookup should be fast. 

I imagine its strictly better if we then forward it after the block lookup. Better to potentially send late, than never send at all.

**Comment by potuz** [comment]
Ever since prysm added a flag to attest as soon as a block is received, in version 1.2.0, their Discord channel has received several complains from users holding many keys.  Since when they propose a block, typically they will be attesting on the same slot, and some of the attestations go missing. This is presumably due to the fact that the attestation leaves the server right after the block, but peers may get the attestation before the block. Apparently Nimbus and Lighthouse do not implement this yet (I do not know about Teku) so if all peers in the subnet are from these clients (or even perhaps just the aggregators), their attestations will be missed. Nimbus adds a delay to attestation propagation when they propose on the same slot. I don't think this is the right approach though.

**Comment by AgeManning** [comment]
Yeah. I think its fine to send the attestation with the block. Once this issue is resolved, we would cache the attestation, search for the block via the RPC and re-process the attestation. In this sense, it should be handled gracefully, provided the block is available from a peer that forwarded it.

**Comment by paulhauner** [comment]
Resolved in #2319
